src/dotty/tools/dotc/reporting/diagnostic/messages.scala

```diff
@@ -729,4 +729,11 @@ object messages {
            | - auxiliary constructors specify the implicit value
            |"""
   }
+
+  case class NoClassOrObjectAfterCase()(implicit ctx: Context)
+  extends Message(27) {
+    val kind = "Syntax"
+    val msg = hl"Only ${"class"} or ${"object"} are allowed after ${"case"}."
+    val explanation = ""
```

Comment:
This is not about `case trait` only. You could write `case val`, `case match` or `case foo`. So in general I think explanation, that only class/object are allowed after case keyword is sufficient. Or we can match token after `case` and if it is `trait`, then print another error message which will inform only about alternatives to `case trait`.